### PR TITLE
Add show in template feature

### PR DESCRIPTION
**Pull request by wader**
Selects the deepest template node for a given position.

### CODE REVIEW

File: app/sources/HFBinaryTemplateController.h [hunk 1]
```diff
@@ -18,6 +18,10 @@ NS_ASSUME_NONNULL_BEGIN
 
 - (void)anchorTo:(NSUInteger)position;
 
+- (void)gotoInTemplateAt:(NSUInteger)position;
+
+- (BOOL)hasTemplate;
```
Comment (kainjow) [inline app/sources/HFBinaryTemplateController.h:23]: Use a property.
```suggestion
@property (readonly) BOOL hasTemplate;
```

File: app/sources/HFBinaryTemplateController.m [hunk 2]
```diff
@@ -382,6 +382,36 @@ - (void)anchorTo:(NSUInteger)position {
     [self rerunTemplate];
 }
 
+- (HFTemplateNode *)findAndExpandDeepestNodeForPosition:(NSUInteger)position startAt:(HFTemplateNode *)node {
+    if (node.children == nil) {
+        return node;
+    }
+    
+    for (HFTemplateNode *c in node.children) {
```
Comment (kainjow) [inline app/sources/HFBinaryTemplateController.m:390]: Don't use a single letter variable. Maybe `childNode`?

File: app/sources/HFBinaryTemplateController.m [hunk 3]
```diff
@@ -382,6 +382,36 @@ - (void)anchorTo:(NSUInteger)position {
...
+        return;
+    }
+    
+    HFTemplateNode *deepest = [self findAndExpandDeepestNodeForPosition:position startAt:self.node];
+    NSInteger itemIndex = [self.outlineView rowForItem:deepest];
+    if (itemIndex < 0) {
+        return;
+    }
+        
+    [self.outlineView selectRowIndexes: [NSIndexSet indexSetWithIndex:itemIndex] byExtendingSelection:NO];
```
Comment (kainjow) [inline app/sources/HFBinaryTemplateController.m:411]: Remove space after selectRowIndexes

File: app/sources/HFBinaryTemplateRepresenter.m [hunk 4]
```diff
@@ -49,16 +49,30 @@ - (void)representerTextView:(HFRepresenterTextView * __unused)sender menu:(NSMen
...
-    [menu addItem:menuItem];
+    
+    NSString *anchorTitle = [NSString stringWithFormat:NSLocalizedString(@"Anchor Template at Offset %llu", nil), self.menuPosition];
+    NSMenuItem *anchorMenuItem = [[NSMenuItem alloc] initWithTitle:anchorTitle action:@selector(anchorTemplatesAt:) keyEquivalent:@""];
+    anchorMenuItem.target = self;
+    anchorMenuItem.enabled = position < self.controller.contentsLength;
+    [menu addItem:anchorMenuItem];
+    
+    if ([self.viewController hasTemplate]) {
+        NSString *gotoInTemplateTitle = NSLocalizedString(@"Goto in template", nil);
```
Comment (kainjow) [inline app/sources/HFBinaryTemplateRepresenter.m:62]: "template" should be capitalized.

I think maybe a different name would be more clear. I'll think about it some more. Maybe something like "Show in Template" ... or "Show Matching Template Item"...

File: app/sources/HFBinaryTemplateRepresenter.m [hunk 5]
```diff
@@ -49,16 +49,30 @@ - (void)representerTextView:(HFRepresenterTextView * __unused)sender menu:(NSMen
...
-    menuItem.enabled = position < self.controller.contentsLength;
-    [menu addItem:menuItem];
+    
+    NSString *anchorTitle = [NSString stringWithFormat:NSLocalizedString(@"Anchor Template at Offset %llu", nil), self.menuPosition];
+    NSMenuItem *anchorMenuItem = [[NSMenuItem alloc] initWithTitle:anchorTitle action:@selector(anchorTemplatesAt:) keyEquivalent:@""];
+    anchorMenuItem.target = self;
+    anchorMenuItem.enabled = position < self.controller.contentsLength;
+    [menu addItem:anchorMenuItem];
+    
+    if ([self.viewController hasTemplate]) {
```
Comment (kainjow) [inline app/sources/HFBinaryTemplateRepresenter.m:61]: Can use property syntax.
```suggestion
    if (self.viewController.hasTemplate) {
```

**Comment by kainjow** [comment]
Thanks! This will be useful.

File: app/sources/HFBinaryTemplateController.h (hunk 1 above)
Comment (wader) [reply app/sources/HFBinaryTemplateController.h]: Thanks, my objective-c i rusty :)

File: app/sources/HFBinaryTemplateRepresenter.m (hunk 4 above)
Comment (wader) [reply app/sources/HFBinaryTemplateRepresenter.m]: Changed to "Show in Template" now. Not sure if the item should be disabled instead or hidden?

**Comment by wader** [comment]
Does it work correctly with template anchoring? it seem so... but probably buy luck

Ok to use NSUInteger for position ? it seems to be 64 bit in 64 bit applications. The objective-c standard library seems to use unsigned long long for file offset and things. Nitpick maybe.

BTW if have bunch of smaller template feature requests should i file issues for each of them or collect into one and maybe split later? have some notes that i can cleanup from when i spend time writing templates.

**Comment by wader** [comment]
Great program btw and templates are really useful. I use it a lot while debugging and writing muxers and demuxers for various media formats :)

**Comment by wader** [comment]
Was reading the suggestions in https://github.com/ridiculousfish/HexFiend/issues/193 ... maybe the selection or some kind of marker should always try to be in "sync" between the hex, ascii and template tree?

**Comment by kumowoon1025** [comment]
> I use it a lot while debugging and writing muxers and demuxers for various media formats :)

That's exactly what I was using it for! I'd be tripping all over the tcl writing these templates and then bam you would open a PR with a much more thorough implementation 😁 

I do feel like "muxing and demuxing" is kind of the limit for a human to handle using a hexadecimal file editor, once I hit the part with all the h264 or hevc or any kind of length coded media sample it may as well be encrypted. Or is that just me? I wonder if there is a codec guru who can picture in their minds the kind of images that are encoded simply by looking the array of sample data bytes!

**Comment by wader** [comment]
Ah glad to hear! hope the templates are useful, they have enabled me to understand lots of things about misbehaving files, if they are truncated, corrupt etc have metadata that is written wrongly etc. Now i feel blind without them :D

Yeah when it gets down to bit level it's a bit hard with a byte oriented hex editor, but wav and flac i can kind of see if it's silence or sinus etc :) I think there were some talks about bit level things in some PR? 

I've done some experiments with a "bit read" api for tcl templates, it works but my tcl implementation of it is very slow so not very usable (you can see it here https://github.com/wader/flac.tcl ... ended up writing a flac decoder while doing it 😄 i have much more detailed flac template also but it timeouts out nearly all flac files).

Also i have been working on decoding api/framework in golang that is bit oriented for a while that i hope would make bit level inspection easier. But no plans to become a editor, mostly just output in some text format like JSON, nice formatted text etc, maybe some kind of viewer UI later. Want to be able to do things like this `fq -f json file.mp4 | jq '.. | .sidx? | select(.)'`.

I actually have tcl implementation of part of the hexfiend template api that can output to json etc, i should clean it up and push it to github.

Nice to talk to ppl who are into messing with binary files!

**Comment by wader** [comment]
https://github.com/wader/textfiend

**Comment by kainjow** [comment]
Sorry for the delay. Can you resolve the conflicts?

**Comment by wader** [comment]
No problem. I've rebased and pushed on hopefully fixed the conflicts correctly. But i have tested it yet as there are issues building for me. Xcode complains that it can't the find Sparkle.framework and i did submodule init/updated with no luck. I tried to re-add it but now i get other errors about missing source files :( hmm. Any ideas?

Do you want to merge the PR as it is and maybe later work on improving with the things i mentioned above about cursor in hex/ascii view should highlight in template tree?

**Comment by kainjow** [comment]
Sparkle is referenced by a submodule, so you'll need to make sure that's setup in git. something like `git submodule update --init`. I use [Fork](https://fork.dev) for most git management though.

**Pull request merged**